### PR TITLE
[MISC] Restructure the help_page printing to a CRTP class. 

### PR DESCRIPTION
**Pull request by smehringer**
Fixes also synopsis printing of html export.
Makes it way easier to keep the export formats synchronized.

### CODE REVIEW

File: include/seqan3/argument_parser/detail/format_base.hpp [hunk 1]
```diff
@@ -196,4 +195,257 @@ class format_base
...
+ */
+template <typename derived_type>
+class format_help_base : public format_base
+{
+private:
+    /*!\name Constructors, destructor and assignment
+     * \brief The (default) constructors.
+     * \{
+     */
+    format_help_base() = default;
```
Comment (eseiler) [inline include/seqan3/argument_parser/detail/format_base.hpp]: Don't you want to document the constructors? I have no idea what's happening here!

File: include/seqan3/argument_parser/detail/format_base.hpp [hunk 2]
```diff
@@ -196,4 +195,257 @@ class format_base
...
+    {}
+
+    //!\brief The destructor.
+    ~format_help_base() = default;
+    //!\}
+
+public:
+    /*!\brief Adds a seqan3::print_list_item call to be evaluated later on.
+     *
+     * \tparam option_type The type of variable in which to store the given command line argument.
```
Comment (eseiler) [inline include/seqan3/argument_parser/detail/format_base.hpp]:
```suggestion
     * \tparam option_type    The type of variable in which to store the given command line argument.
```

File: include/seqan3/argument_parser/detail/format_base.hpp [hunk 3]
```diff
@@ -196,4 +195,257 @@ class format_base
...
+     */
+    void parse(argument_parser_meta_data & parser_meta)
+    {
+        meta = parser_meta;
+
+        static_cast<derived_type &>(*this).print_header();
+
+        if (!meta.synopsis.empty())
+        {
+            static_cast<derived_type &>(*this).print_section("Synopsis");
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_base.hpp]: You could make a private function that casts this for you. `derived` maybe

File: include/seqan3/argument_parser/detail/format_help.hpp [hunk 4]
```diff
@@ -37,8 +37,14 @@ namespace seqan3::detail
...
  * and only evaluated when calling format_help::parse().
  */
-class format_help : public format_base
+class format_help : public format_help_base<format_help>
 {
+    //!\brief The CRTP base class type.
+    using base_type = format_help_base<format_help>;
+
+    //!\brief Befriend the base class to give access to the private member functions.
+    friend format_help_base;
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_help.hpp]:
```suggestion
    friend base_type
```

File: include/seqan3/argument_parser/detail/format_html.hpp [hunk 5]
```diff
@@ -31,166 +31,30 @@ namespace seqan3::detail
...
-     */
-    void add_line(std::string const & text, bool const line_is_paragraph)
-    {
-        parser_set_up_calls.push_back([this, text, line_is_paragraph] ()
-        {
-            print_line(text, line_is_paragraph);
-        });
-    }
+    //!\brief Befriend the base class to give access to the private member functions.
+    friend format_help_base;
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_html.hpp]:
```suggestion
    friend base_type;
```

File: include/seqan3/argument_parser/detail/format_html.hpp [hunk 6]
```diff
@@ -31,166 +31,30 @@ namespace seqan3::detail
...
+    format_html() = default;
+    format_html(format_html const & pf) = default;
+    format_html & operator=(format_html const & pf) = default;
+    format_html(format_html &&) = default;
+    format_html & operator=(format_html &&) = default;
+    ~format_html() = default;
+    //!\}
+
+    //!\copydoc format_help_base(bool)
+    format_html(bool advanced) : base_type{advanced}
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_html.hpp]:
```suggestion
    format_html(bool const advanced) : base_type{advanced}
```

File: include/seqan3/argument_parser/detail/format_man.hpp [hunk 7]
```diff
@@ -32,160 +32,30 @@ namespace seqan3::detail
...
-
-        positional_option_calls.push_back([this, &value, desc, validator] ()
-        {
-            std::string key{"\\fBARGUMENT-" + std::to_string(positional_option_count) +
-                            "\\fP " + option_type_and_list_info(value)};
-            print_list_item(key, (desc + " " + validator.get_help_page_message()));
-        });
-    }
+    //!\brief Befriend the base class to give access to the private member functions.
+    friend format_help_base;
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_man.hpp]:
```suggestion
    friend base_type;
```

File: include/seqan3/argument_parser/detail/format_man.hpp [hunk 8]
```diff
@@ -32,160 +32,30 @@ namespace seqan3::detail
...
+    format_man() = default;
+    format_man(format_man const & pf) = default;
+    format_man & operator=(format_man const & pf) = default;
+    format_man(format_man &&) = default;
+    format_man & operator=(format_man &&) = default;
+    ~format_man() = default;
+    //!\}
+
+    //!\copydoc format_help_base(bool)
+    format_man(bool advanced) : base_type{advanced}
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_man.hpp]:
```suggestion
    format_man(bool const advanced) : base_type{advanced}
```

File: test/unit/argument_parser/detail/format_html_test.cpp [hunk 9]
```diff
@@ -29,25 +29,23 @@ TEST(html_test, html)
...
-                           "<h1>empty_options</h1>"
-                           "<div></div>"
-                           "<h2>Synopsis</h2>"
-                           "<h2>Version</h2>"
-                           "<strong>Last update:</strong> <br>"
-                           "<strong>empty_options version:</strong> <br>"
-                           "<strong>SeqAn version:</strong> 3.0.0<br>"
-                           "<br>"
+    std_out = testing::internal::GetCapturedStdout();
+    expected = std::string("<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.01//EN\" http://www.w3.org/TR/html4/strict.dtd\">\n"
```
Comment (rrahn) [inline test/unit/argument_parser/detail/format_html_test.cpp]: Would be nicer if we could use rapidxml somehow to test the DOM tree. It is not very important but perhaps you can open a card for this?

Comment (smehringer) [reply test/unit/argument_parser/detail/format_html_test.cpp]: I added a card

File: include/seqan3/argument_parser/detail/format_help.hpp [hunk 10]
```diff
@@ -37,212 +37,29 @@ namespace seqan3::detail
...
-     * The member variable itself is filled when copied over from the argument_parser
-     * when calling format_parse::parse. That way all the information needed are
-     * there, when the actual printing starts.
-     *
-     * This function is not private because it is needed for short but nicely
-     * formatted (error) output to the command line.
-     */
-    argument_parser_meta_data meta;
+    //!\copydoc format_help_base(bool)
+    format_help(bool const advanced) : base_type{advanced}
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_help.hpp]: it looks like this constructor is not in the doxygen group for constructors anymore? 🤔

File: include/seqan3/argument_parser/detail/format_html.hpp [hunk 11]
```diff
@@ -31,166 +31,29 @@ namespace seqan3::detail
...
+    format_html() = default;                                   //!< Defaulted.
+    format_html(format_html const & pf) = default;             //!< Defaulted.
+    format_html & operator=(format_html const & pf) = default; //!< Defaulted.
+    format_html(format_html &&) = default;                     //!< Defaulted.
+    format_html & operator=(format_html &&) = default;         //!< Defaulted.
+    ~format_html() = default;                                  //!< Defaulted.
+    //!\}
+
+    //!\copydoc format_help_base(bool)
+    format_html(bool const advanced) : base_type{advanced}
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_html.hpp]: Not in the doxygen group for constructors anymore.

File: include/seqan3/argument_parser/detail/format_man.hpp [hunk 12]
```diff
@@ -32,160 +32,29 @@ namespace seqan3::detail
...
+    format_man() = default;                                  //!< Defaulted.
+    format_man(format_man const & pf) = default;             //!< Defaulted.
+    format_man & operator=(format_man const & pf) = default; //!< Defaulted.
+    format_man(format_man &&) = default;                     //!< Defaulted.
+    format_man & operator=(format_man &&) = default;         //!< Defaulted.
+    ~format_man() = default;                                 //!< Defaulted.
+    //!\}
+
+    //!\copydoc format_help_base(bool)
+    format_man(bool const advanced) : base_type{advanced}
```
Comment (rrahn) [inline include/seqan3/argument_parser/detail/format_man.hpp]: same here

**Comment by smehringer** [comment]
@rrahn If you approve, you can merge it.

**Comment by smehringer** [comment]
@rrahn ping :)

**Pull request merged**